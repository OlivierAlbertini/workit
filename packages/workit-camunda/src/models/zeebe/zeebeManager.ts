/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { inject, injectable, named } from 'inversify';
import { IWorkflowClient } from 'workit-types';
import { SERVICE_IDENTIFIER } from '../../config/constants/identifiers';
import { TAG } from '../../config/constants/tag';
import { ClientManager } from '../camunda-n-mq/clientManager';

@injectable()
export class ZeebeManager extends ClientManager<IWorkflowClient> {
  constructor(@inject(SERVICE_IDENTIFIER.camunda_client) @named(TAG.zeebe) client: IWorkflowClient) {
    super(client);
  }
}
