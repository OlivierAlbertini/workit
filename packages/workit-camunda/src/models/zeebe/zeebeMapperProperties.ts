/*!
 * Copyright (c) 2019 Ville de Montreal. All rights reserved.
 * Licensed under the MIT license.
 * See LICENSE file in the project root for full license information.
 */

import { IProperties } from '../camunda-n-mq/specs/properties';
import { IEmptyPayload, IPayload } from './specs/payload';

export class ZeebeMapperProperties {
  public static map<TVariables = any, TProps = any>(obj: IPayload<TVariables, TProps>): IProperties<TProps> {
    if (!obj.variables) {
      obj.variables = {} as any;
    }
    return {
      businessKey: (obj.variables as any).businessKey, // retro compatibility with bpmn workflow engine
      activityId: obj.elementId,
      processInstanceId: obj.elementInstanceKey,
      workflowDefinitionVersion: obj.workflowDefinitionVersion,
      workflowInstanceKey: obj.workflowInstanceKey,
      workflowKey: obj.workflowKey,
      bpmnProcessId: obj.bpmnProcessId,
      customHeaders: { ...obj.customHeaders },
      jobKey: obj.key,
      retries: obj.retries,
      topicName: obj.type,
      workerId: obj.worker,
      lockExpirationTime: new Date(Number(obj.deadline))
    };
  }

  public static unmap<TProps = any>(props: IProperties<TProps>): IEmptyPayload<TProps> {
    return {
      elementId: props.activityId,
      elementInstanceKey: props.processInstanceId,
      workflowKey: props.workflowKey,
      workflowInstanceKey: props.workflowInstanceKey,
      workflowDefinitionVersion: props.workflowDefinitionVersion,
      bpmnProcessId: props.bpmnProcessId,
      retries: props.retries,
      worker: props.workerId,
      deadline: props.lockExpirationTime.getTime().toString(),
      customHeaders: { ...props.customHeaders },
      key: props.jobKey,
      type: props.topicName
    };
  }
}
