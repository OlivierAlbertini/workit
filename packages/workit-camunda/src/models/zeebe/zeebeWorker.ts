import { inject, injectable, named } from 'inversify';
import { SERVICE_IDENTIFIER } from '../../config/constants/identifiers';
import { TAG } from '../../config/constants/tag';
import { Client } from '../camunda-n-mq/client';
import { IClient } from '../camunda-n-mq/specs/client';
import { IProcessHandler } from '../core/processHandler/specs/processHandler';
import { Worker } from '../core/worker';

@injectable()
export class ZeebeWorker extends Worker {
  constructor(
    @inject(SERVICE_IDENTIFIER.camunda_client) @named(TAG.zeebe) client: Client<IClient>,
    @inject(SERVICE_IDENTIFIER.success_strategy) processHandler: IProcessHandler
  ) {
    super(client, processHandler);
  }
}
