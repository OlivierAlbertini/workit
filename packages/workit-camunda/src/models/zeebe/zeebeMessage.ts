// Copyright (c) Ville de Montreal. All rights reserved.
// Licensed under the MIT license.
// See LICENSE file in the project root for full license information.

import { ZBClient } from 'zeebe-node';
import { getVariablesWhenChanged } from '../../utils/utils';
import { ICamundaService } from '../camunda-n-mq/specs/camundaService';
import { FailureException } from '../camunda-n-mq/specs/failureException';
import { IMessage } from '../camunda-n-mq/specs/message';
import { IProperties } from '../camunda-n-mq/specs/properties';
import { CamundaClientTracer } from '../core/instrumentations/camundaClientTracer';
import { APM } from '../core/instrumentations/enums/apm';
import { ICCInstrumentationHandler } from '../core/instrumentations/specs/instrumentation';
// import { ProxyFactory } from '../core/proxyFactory';
import { IPayload } from './specs/payload';
import { ZeebeMapperProperties } from './zeebeMapperProperties';

export class ZeebeMessage {
  public static wrap(
    payload: IPayload,
    complete,
    client: ZBClient,
    apm: ICCInstrumentationHandler
  ): [IMessage, ICamundaService] {
    const properties = ZeebeMapperProperties.map(payload);
    const tracer = apm.get(APM.tracer) as CamundaClientTracer;
    const messageWithoutSpan = {
      body: payload.payload,
      properties: properties as IProperties
    };
    const spans = tracer.createRootSpanOnMessage(messageWithoutSpan);
    // const messageWithProxy = ProxyFactory.create({
    //   body: messageWithoutSpan.body,
    //   properties: messageWithoutSpan.properties,
    //   span,
    // });
    return [
      {
        body: messageWithoutSpan.body,
        properties: messageWithoutSpan.properties,
        spans
      },
      {
        hasBeenThreated: false,
        async ack(message: IMessage) {
          if (this.hasBeenThreated) {
            return;
          }
          const vars = getVariablesWhenChanged(message, msg => ZeebeMessage.unwrap(msg));
          complete(vars);
          this.hasBeenThreated = true;
        },
        async nack(error: FailureException) {
          if (this.hasBeenThreated) {
            return;
          }
          const retries = error.retries;
          await client.failJob({ retries, jobKey: payload.key, errorMessage: error.message });
          this.hasBeenThreated = true;
        }
      }
    ];
  }

  /**
   * Shallow copy
   */
  public static unwrap(message: IMessage): IPayload {
    const emptyPayload = ZeebeMapperProperties.unmap(message.properties);
    emptyPayload.payload = message.body;
    return emptyPayload as IPayload;
  }
}
