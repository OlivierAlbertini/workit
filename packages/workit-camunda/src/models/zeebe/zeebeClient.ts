// Copyright (c) Ville de Montreal. All rights reserved.
// Licensed under the MIT license.
// See LICENSE file in the project root for full license information.

import { optional } from 'inversify';
import { ZBClient, ZBWorker } from 'zeebe-node';
import { ICamundaService } from '../camunda-n-mq/specs/camundaService';
import { IClient } from '../camunda-n-mq/specs/client';
import { ICreateWorkflowInstance } from '../camunda-n-mq/specs/createWorkflowInstance';
import { ICreateWorkflowInstanceResponse } from '../camunda-n-mq/specs/createWorkflowInstanceResponse';
import { IDeployWorkflowResponse } from '../camunda-n-mq/specs/deployWorkflowResponse';
import { IMessage } from '../camunda-n-mq/specs/message';
import { IPublishMessage } from '../camunda-n-mq/specs/publishMessage';
import { IUpdateWorkflowRetry } from '../camunda-n-mq/specs/updateWorkflowRetry';
import { IUpdateWorkflowVariables } from '../camunda-n-mq/specs/updateWorkflowVariables';
import { IWorkflowClient } from '../camunda-n-mq/specs/workflowClient';
import { IWorkflowDefinition, IWorkflowDefinitionRequest } from '../camunda-n-mq/specs/workflowDefinition';
import { ICCInstrumentationHandler } from '../core/instrumentations/specs/instrumentation';
import { IPayload } from './specs/payload';
import { IZeebeOptions } from './specs/zeebeOptions';
import { ZeebeMessage } from './zeebeMessage';
export interface IZeebeClient {
  [custom: string]: any;
}

export class ZeebeClient implements IClient, IWorkflowClient {
  private readonly _client: ZBClient;
  private _worker: ZBWorker | undefined;
  private readonly _config: IZeebeOptions;
  private readonly _apm: ICCInstrumentationHandler;

  constructor(config: IZeebeOptions, apm: ICCInstrumentationHandler, @optional() client?: ZBClient) {
    this._client = client || new ZBClient(config.baseUrl, config);
    this._apm = apm;
    this._config = config;
  }
  public subscribe(onMessageReceived: (message: IMessage, service: ICamundaService) => Promise<void>): Promise<void> {
    this._worker = this._client.createWorker(
      this._config.workerId || 'some-random-id',
      this._config.topicName,
      async (payload: IPayload, complete: (content: IPayload) => void) => {
        const [message, service] = ZeebeMessage.wrap(payload, complete, this._client, this._apm);
        try {
          await onMessageReceived(message, service);
        } catch (error) {
          this._apm.onMessageFailed(error, message);
        } finally {
          this._apm.onMessageSuccess(message);
        }
      },
      this._config
    );
    return Promise.resolve();
  }
  public async deployWorkflow(bpmnPath: string): Promise<IDeployWorkflowResponse> {
    const result = await this._client.deployWorkflow(bpmnPath);
    return {
      workflows: result.workflows,
      key: result.key.toString() // TODO: interface say number but it return string, need to PR to zeebe-node
    };
  }
  public async getWorkflows(): Promise<IWorkflowResponse> {
    return this._client.listWorkflows();
  }
  public async getWorkflow(payload: IWorkflowDefinitionRequest): Promise<IWorkflowDefinition> {
    return this._client.getWorkflow(payload);
  }
  public updateVariables<T = any>(model: IUpdateWorkflowVariables<T>): Promise<void> {
    return this._client.updateWorkflowInstancePayload({
      elementInstanceKey: model.processInstanceId,
      payload: model.variables
    });
  }
  public updateJobRetries(payload: IUpdateWorkflowRetry): Promise<void> {
    return this._client.updateJobRetries(payload);
  }
  /**
   * Publish a message event
   * If you don't specify correlationKey, message will be treated as message start event
   */
  public publishMessage<T>(payload: IPublishMessage<T, string>): Promise<void> {
    return this._client.publishMessage({
      correlationKey: payload.correlation || '__MESSAGE_START_EVENT__',
      payload: payload.variables,
      messageId: payload.messageId,
      timeToLive: payload.timeToLive,
      name: payload.name
    });
  }
  public createWorkflowInstance<T>(model: ICreateWorkflowInstance<T>): Promise<ICreateWorkflowInstanceResponse> {
    return this._client.createWorkflowInstance(model.bpmnProcessId, model.variables, model.version);
  }
  public cancelWorkflowInstance(instanceId: string): Promise<void> {
    return this._client.cancelWorkflowInstance(instanceId);
  }
  public resolveIncident(incidentKey: string): Promise<void> {
    return this._client.resolveIncident(incidentKey);
  }
  public unsubscribe(): Promise<void> {
    if (this._worker) {
      return this._worker.close();
    }
    return Promise.resolve();
  }
}
