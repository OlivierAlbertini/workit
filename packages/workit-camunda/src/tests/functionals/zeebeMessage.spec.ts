// Copyright (c) Ville de Montreal. All rights reserved.
// Licensed under the MIT license.
// See LICENSE file in the project root for full license information.

import * as opentracing from 'opentracing';
import { ZeebeMessage } from '../../models/zeebe/zeebeMessage';

// tslint:disable:ter-prefer-arrow-callback
// tslint:disable:only-arrow-functions
// tslint:disable:max-func-body-length
describe('zeebeMessage', function() {
  it.only('unmap', () => {
    const message = {
      body: { hello: 'world' },
      properties: {
        workflowInstanceKey: '38963',
        bpmnProcessId: 'test-process',
        workflowDefinitionVersion: 4,
        workflowKey: '8806',
        activityId: 'ServiceTask_0xdwuw7',
        elementInstanceKey: '38967',
        customHeaders: { jwt: 'jwt fake' },
        jobKey: '38968',
        processInstanceId: '38963',
        retries: 1,
        lockExpirationTime: new Date(1562269537659),
        topicName: 'topic_demo',
        workerId: 'demo'
      },
      spans: new opentracing.Span()
    };
    const zeebeObject = ZeebeMessage.unwrap(message);
    expect(zeebeObject).toMatchSnapshot();
  });
});
