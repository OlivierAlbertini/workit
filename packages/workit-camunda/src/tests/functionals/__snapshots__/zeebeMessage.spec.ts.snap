// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zeebeMessage unmap 1`] = `
Object {
  "customHeaders": Object {
    "jwt": "jwt fake",
  },
  "deadline": "1562269537659",
  "jobHeaders": Object {
    "bpmnProcessId": "test-process",
    "elementId": "ServiceTask_0xdwuw7",
    "elementInstanceKey": "38963",
    "workflowDefinitionVersion": 4,
    "workflowInstanceKey": "38963",
    "workflowKey": "8806",
  },
  "key": "38968",
  "retries": 1,
  "type": "topic_demo",
  "variables": Object {
    "hello": "world",
  },
  "worker": "demo",
}
`;
